### ISSUE TITLE
Delay when sending e-mails

### ISSUE THREAD

**Issue by dabide** [hardware/mailserver#312]
#### Classification
* [X] Question

#### Reproducibility
* [X] Always

#### Description
There is always about ten seconds of delay when Thunderbird sends a mail through my SMTP server. It seems to have to do with an RBL lookup, but that is difficult to say. Is this normal?

#### Steps to reproduce
1. Send an e-mail using SMTP and STARTTLS on port 587

#### Expected results
Connect - send - finished in max a second or two.

#### Actual results
Connect - send - finished takes about ten seconds.

#### Debugging information
```
2018-10-03T05:38:56.120861+00:00 mail postfix/submission/smtpd[2700]: connect from ........
2018-10-03T05:38:56.338654+00:00 mail postfix/submission/smtpd[2700]: Anonymous TLS connection established from ........
net[aaa.bbb.ccc.ddd]: TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits)
2018-10-03T05:39:06.636424+00:00 mail postfix/submission/smtpd[2700]: warning: aaa.bb.dbl.spamhaus.org: RBL lookup error: Host or domain
name not found. Name service error for name=aaa.bb.dbl.spamhaus.org type=A: Host not found, try again
2018-10-03T05:39:06.649030+00:00 mail postfix/submission/smtpd[2700]: 9E6113F38: client=........, sas
l_method=PLAIN, sasl_username=user@aaa.bb
2018-10-03T05:39:06.923216+00:00 mail postfix/authclean/cleanup[2703]: 9E6113F38: replace: header Received: from ........ ??(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits))??(No client certificate requested)??by mail. from ........; from=........ to=........ proto=ESMT
P helo=<[bbb.ccc.ddd.eee]>: Received: from authenticated-user (mail.aaa.bb [127.0.0.1])??(using TLSv1.2 with cipher ECDHE-RSA-AES256-GCM-S
HA384 (256/256 bits))??(No client certificate requested)??by mail.aaa.bb (Postfix) with ESMTPSA id 9E6113F38;??Wed,  3 Oct 2018 05:38:56
+0000 (UTC)
```

**Comment by dabide** [comment]
It definitely seems to be connected to the RBL check. With debug logging:

```
2018-10-03T09:24:10.596094+00:00 mail postfix/submission/smtpd[757]: reject_rbl_domain: dbl.spamhaus.org user@aaa.bb
2018-10-03T09:24:20.606613+00:00 mail postfix/submission/smtpd[757]: dns_query: aaa.bb.dbl.spamhaus.org (A): Host not found, try again
2018-10-03T09:24:20.606643+00:00 mail postfix/submission/smtpd[757]: warning: aaa.bb.dbl.spamhaus.org: RBL lookup error: Host or domain name not found. Name service error for name=aaa.bb.dbl.spamhaus.org type=A: Host not found, try again
```

**Comment by denji** [comment]
Are you running systemd-resolved on host machine?

**Comment by denji** [comment]
I am using systemd-resolved with dnssec enabled and Unbound disabled (`DISABLE_DNS_RESOLVER=true`).

> https://www.freedesktop.org/software/systemd/man/resolved.conf.html
> `DNS=8.8.8.8 8.8.4.4` / `DNS=1.1.1.1 1.0.0.1`;
> Cache=yes
> DNSSEC=allow-downgrade

If there was a Knot Resolver, I'd to use it too.

**Comment by dabide** [comment]
Disabling unbound solves the problem. Are there any logs that could tell why?

**Comment by hardware** [comment]
Your hosting provider probably apply a rate limit on your DNS rbl queries.

> Are there any logs that could tell why?

Enable unbound and after few days, check unbound stats with :

```
docker exec -ti mailserver unbound-control stats_noreset
```

if **recursivereplies** is very far away from **cachemiss** value, you have a problem.

Example with 40% requests fail :

```
thread0.num.queries=422
thread0.num.cachehits=0
thread0.num.cachemiss=422
thread0.num.recursivereplies=267
```

Example with 0% request fail :

```
thread0.num.queries=633
thread0.num.cachehits=333
thread0.num.cachemiss=300
thread0.num.recursivereplies=300
```

**Comment by hardware** [comment]
I will add a proper Debug mode for unbound with `use-syslog` when `DEBUG_MODE` is enabled.

For the moment, you can specify a file here :
https://github.com/hardware/mailserver/blob/9d05590c37f4ede0295ede486eb89d3e9385a137/rootfs/etc/unbound/unbound.conf#L4

And increase verbosity level.

https://www.nlnetlabs.nl/documentation/unbound/unbound.conf/

**Comment by denji** [comment]
We disable under prodserv unbound `DISABLE_DNS_RESOLVER=true`.
Knot Resolver works well, systemd-resolved also works fine in host mode.

**Comment by hardware** [comment]
Keep in mind that if you switch `DISABLE_DNS_RESOLVER` to `true`, DNSSEC is disabled in Postfix and Rspamd configuration.

https://github.com/hardware/mailserver/blob/9d05590c37f4ede0295ede486eb89d3e9385a137/rootfs/usr/local/bin/run.sh#L466-L476

The only reason to use this option is when your provider apply a rate limiting to your DNS queries, and rspamd actually does a lot.

**Comment by hardware** [comment]
@dabide : I added unbound in debug mode (only available in `1.1-latest` for now), please test with this mode and post some logs and unbound-control stats here.

```yml
mailserver:
  image: hardware/mailserver:1.1-latest
  environment:
    - DISABLE_DNS_RESOLVER=false
    - DEBUG_MODE=unbound
```
```bash
docker logs -f mailserver
docker exec mailserver unbound-control stats_noreset
```

**Comment by denji** [comment]
@hardware [`DNSSEC=allow-downgrade`](https://www.freedesktop.org/software/systemd/man/resolved.conf.html#DNSSEC=)
